crates/sui-core/src/authority/authority_store_types.rs

```diff
@@ -62,47 +64,50 @@ where
 }
 
 impl StoreMoveObject {
-    pub fn digest(&self) -> ObjectDigest {
-        ObjectDigest::new(sha3_hash(self))
+    pub fn digest(&self) -> ObjectContentDigest {
+        // expected to be called on constructed object with default ref count 1
+        assert_eq!(self.ref_count, 1);
+        ObjectContentDigest::new(sha3_hash(self))
     }
 }
 
 pub struct StoreObjectPair(pub StoreObject, pub Option<StoreMoveObject>);
 
-impl From<Object> for StoreObjectPair {
-    fn from(object: Object) -> Self {
-        let mut indirect_object = None;
+pub(crate) fn get_store_object_pair(
+    object: Object,
+    enable_indirect_objects: bool,
+) -> StoreObjectPair {
+    let mut indirect_object = None;
 
-        let data = match object.data {
-            Data::Package(package) => StoreData::Package(package),
-            Data::Move(move_obj) => {
-                // TODO: add real heuristic to decide if object needs to be stored indirectly
-                if cfg!(test) {
-                    let move_object = StoreMoveObject {
-                        type_: move_obj.type_.clone(),
-                        has_public_transfer: move_obj.has_public_transfer(),
-                        contents: move_obj.contents().to_vec(),
-                        ref_count: 1,
-                    };
-                    let digest = move_object.digest();
-                    indirect_object = Some(move_object);
-                    StoreData::IndirectObject(IndirectObjectMetadata {
-                        version: move_obj.version(),
-                        digest,
-                    })
-                } else {
-                    StoreData::Move(move_obj)
-                }
+    let data = match object.data {
+        Data::Package(package) => StoreData::Package(package),
+        Data::Move(move_obj) => {
+            // TODO: add real heuristic to decide if object needs to be stored indirectly
+            if !Coin::is_coin(&move_obj.type_) && enable_indirect_objects {
```

Comment:
I ran local benchmark with all available workloads we have so far: coin transfer, shared counter, delegation. I also disabled the pruner for better comparison. 
Indirect object schema doesn't improve storage usage for any of them. It ends up being somewhat comparable + having some overhead.
The reason is that objects are actually either relatively small or getting mutated in every transaction.
Optimization makes sense for relatively large immutable(or rarely changed) objects

I slightly modified the counter example. Now it's immutable and contains an extra field of [u8;1024] 
(https://gist.github.com/phoenix-o/ad5a93dd5f043ea1e4855ce7b1e6add1)
Results are much more noticeable:
old version:
Total num keys = 3849, total key bytes = 153960, total value bytes = 240670

indirect version:
 objects table: Total num keys = 3849, total key bytes = 153960, total value bytes = 621903
 indirect move objects table : Total num keys = 661, total key bytes = 26440, total value bytes = 71619

Reply:
@phoenix-o One thing I don't fully understand is why the `total_value_bytes` is higher (621903 > 240670) for objects table in the indirect version? Shouldn't it be smaller given the content for some of them moved away to the indirect table?

Reply:
oh, it's a result of load test run, so majority of objects are still coins paid for gas fees, not only counters

Reply:
oh ok, so the reason the `total_value_bytes` is higher is because the object metadata we are now storing is more in byte size vs serialized object?

Reply:
oh, sorry for confusion, I probably miscopied output, here are the fresh run results:
old version
   Total num keys = 2723, total key bytes = 108480, total value bytes = 1695301
new version:
   Total num keys = 2723, total key bytes = 108920, total value bytes = 461907
   Total num keys = 492, total key bytes = 19680, total value bytes = 55564

to reiterate, I ran slightly modified version of shared counter workload, where counters are immutable and have extra memory consumption. Code is [here](https://gist.github.com/phoenix-o/ad5a93dd5f043ea1e4855ce7b1e6add1)
This tests creates two type of objects: said counters and normal coins used for transfer to pay for gas fees.
You can check num keys. So with indirection on, indirect move objects table contains only new counters, while objects table contains counter metadata and coin objects. And if you add up total value bytes for both tables it's lower than bytes used in old version

Reply:
cool, this makes a lot more sense. Thanks!